src/main/java/com/aws/greengrass/builtin/services/telemetry/ComponentMetricIPCEventStreamAgent.java

```diff
@@ -0,0 +1,197 @@
+import software.amazon.awssdk.aws.greengrass.model.InvalidArgumentsError;
+import software.amazon.awssdk.aws.greengrass.model.PutComponentMetricRequest;
+import software.amazon.awssdk.aws.greengrass.model.PutComponentMetricResponse;
+import software.amazon.awssdk.aws.greengrass.model.ServiceError;
+import software.amazon.awssdk.aws.greengrass.model.UnauthorizedError;
+import software.amazon.awssdk.eventstreamrpc.OperationContinuationHandlerContext;
+import software.amazon.awssdk.eventstreamrpc.model.EventStreamJsonMessage;
+
+import java.nio.charset.StandardCharsets;
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Pattern;
+import javax.inject.Inject;
+
+import static com.aws.greengrass.ipc.common.ExceptionUtil.translateExceptions;
+import static com.aws.greengrass.ipc.modules.ComponentMetricIPCService.PUT_COMPONENT_METRIC_SERVICE_NAME;
+
+public class ComponentMetricIPCEventStreamAgent {
+    private static final Logger logger = LogManager.getLogger(ComponentMetricIPCEventStreamAgent.class);
+    private static final Pattern SERVICE_NAME_REGEX = Pattern.compile("^aws\\.");
+    private static final String NON_ALPHANUMERIC_REGEX = "[^A-Za-z0-9]";
+    private static final String SERVICE_NAME = "ServiceName";
+
+    @Getter(AccessLevel.PACKAGE)
+    private final Map<String, MetricFactory> metricFactoryMap = new HashMap<>();
+    private final AuthorizationHandler authorizationHandler;
+
+    @Inject
+    ComponentMetricIPCEventStreamAgent(AuthorizationHandler authorizationHandler) {
+        this.authorizationHandler = authorizationHandler;
+    }
+
+    public PutComponentMetricOperationHandler getPutComponentMetricHandler(
+            OperationContinuationHandlerContext context) {
+        return new PutComponentMetricOperationHandler(context);
+    }
+
+    class PutComponentMetricOperationHandler extends GeneratedAbstractPutComponentMetricOperationHandler {
+        private final String serviceName;
+
+        protected PutComponentMetricOperationHandler(OperationContinuationHandlerContext context) {
+            super(context);
+            serviceName = context.getAuthenticationData().getIdentityLabel();
+        }
+
+        @Override
+        protected void onStreamClosed() {
+            // NA
+        }
+
+        @SuppressWarnings({"PMD.PreserveStackTrace", "PMD.AvoidCatchingGenericException"})
+        @Override
+        public PutComponentMetricResponse handleRequest(PutComponentMetricRequest componentMetricRequest) {
+            return translateExceptions(() -> {
+                logger.atDebug().kv(SERVICE_NAME, serviceName)
+                        .log("Received putComponentMetricRequest from component " + serviceName);
+
+                // Authorize request for given operation
+                try {
+                    doAuthorization(this.getOperationModelContext().getOperationName(), serviceName);
```

Comment:
is the operation always going to be "putComponentMetric" here?

Reply:
It would always be `aws.greengrass#PutComponentMetric`